pennylane_lightning/core/src/lightning_qubit.toml

```diff
@@ -0,0 +1,132 @@
+schema = 1
+
+[device]
+name = "lightning.qubit"
+precision = ['float32', 'float64']
+
+[operations]
+observables = [
+        "PauliX",
+        "PauliY",
+        "PauliZ",
+        "Hadamard",
+        "Hermitian",
+        "Identity",
+        "Projector",
+        "SparseHamiltonian",
+        "Hamiltonian",
+        "Sum",
+        "SProd",
+        "Prod",
+        "Exp",
+]
+
+# These are gates that do not need decomposition.
+# Also gates which are natively supported by the device.
+[[operations.gates]]
+native = [
+        "Identity",
+        "PauliX",
+        "PauliY",
+        "PauliZ",
+        "Hadamard",
+        "S",
+        "T",
+        "PhaseShift",
+        "RX",
+        "RY",
+        "RZ",
+        "Rot",
+        "CNOT",
+        "CY",
+        "CZ",
+        "SWAP",
+        "CSWAP",
+        "Toffoli",
+        "IsingXX",
+        "IsingXY",
+        "IsingYY",
+        "IsingZZ",
+        "ControlledPhaseShift",
+        "CRX",
+        "CRY",
+        "CRZ",
+        "CRot",
+        "SingleExcitation",
+        "SingleExcitationPlus",
+        "SingleExcitationMinus",
+        "DoubleExcitation",
+        "DoubleExcitationPlus",
+        "DoubleExcitationMinus",
+        "MultiRZ",
+	# QubitUnitary is technically
+	# "applyMatrix", but it doesn't
+	# belong in matrix gates.
+        "QubitUnitary",
+]
+
+# Gates listed in qml.device.ops
+# which the authors of this file would like to be decomposed
+# according to rules outlined in the decomposition strategy for each
+# operation.
+decomp = [
+	# All adjoint operations in lightning_qubit
+	# are applied via applyMatrix.
+        "Adjoint(S)",
+        "Adjoint(T)",
+        "Adjoint(SX)",
+        "Adjoint(ISWAP)",
+        "Adjoint(SISWAP)",
+
+	# The rest.
+        "BasisState",
+        "ECR",
+        "QubitStateVector",
+        "StatePrep",
+        "ControlledQubitUnitary",
+        "DiagonalQubitUnitary",
+        "SX",
+        "ISWAP",
+        "PSWAP",
+        "SISWAP",
+        "SQISW",
+        "CPhase",
+        "QubitCarry",
+        "QubitSum",
+        "OrbitalRotation",
+        "QFT",
+]
+
+# Gates listed in qml.device.ops
+# which the authors of this file would like to be decomposed into QubitUnitary.
+matrix = [
+        "MultiControlledX",
```

Comment:
Maybe nothing to do here for now, but #516 has now been merged. `applyOperation` is overloaded and accepts an extra argument `controlled_wires`. All gates listed in `native` (expect the controlled gates) can now be arbitrarily controlled. The control values are assumed to be `1`s, so if there are `0`s the gate should be decomposed with PauliX sandwiches. So `MultiControlledX` for example can be applied passing `"PauliX"` to `applyOperation` along with `controlled_wires`.

Reply:
I can move MultiControlledX to native if you prefer that.

Reply:
That's an interesting point, but I'm not sure that would help us directly 🤔 If there is no direct kernel support, the gate would still need to be transformed into a different gate (`C(PauliX)`) somehow, and the PennyLane compiler probably shouldn't be aware of such gate equivalences (the decomposition or gate set mapping framework is responsible for that).

Reply:
There are a few "special" gates like that for which `qml.ctrl(op)` is not named `C(op.name)`. On top of my head: `MultiControlledX`, `Toffoli`, `ControlledQubitUnitary`, `CNOT`, `CSWAP`, `CY`, `CZ`, `ControlledPhaseShift`, `CRX`, `CRY`, `CRZ`, `CRot`. This actually depends on the number of controlled wires for most of them.

Reply:
True, I think most of these have native kernel support though:
https://github.com/PennyLaneAI/pennylane-lightning/blob/master/pennylane_lightning/core/src/gates/GateOperation.hpp#L27-L70

The exception being `MultiControlledX` and `ControlledQubitUnitary`. I think in that vein it would be nice if `MultiControlledX` internally (C++ layer) redirected to the controlled application of the `PauliX` kernel, if that currently happens in the Python device.
Regarding `ControlledQubitUnitary`, does the recent implementation of controlled gates in lightning include the capability of a controlled unitary matrix, or was that only implemented for named gates?

Reply:
> EDIT: just noticed that this strategy only applies if all control wires are tested for `1`, in this case I would say decomposition is still the best strategy for now unless lightning adds a kernel that can handle different control values

We plan to pass control values in a follow PR, but this is not possible at the moment. 

> Regarding `ControlledQubitUnitary`, does the recent implementation of controlled gates in lightning include the capability of a controlled unitary matrix, or was that only implemented for named gates?

This is implemented for named gates, matrices and generators (all with the limitations of `1`-controlled wires though).

Reply:
Ah single control wire only? Good to know, thanks!

Reply:
No sorry, you can have an arbitrary number of control wires, but the control values are uniformly `1`. That's what I meant by `1`-controlled.

Reply:
I will be closing this discussion, feel free to reopen it if you believe you have more things to discuss.

Reply:
Just posted a comment on this that is actually a continuation of the discussion here:
https://github.com/PennyLaneAI/pennylane-lightning/pull/566#issuecomment-1852818852